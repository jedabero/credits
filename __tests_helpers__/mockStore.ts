// eslint-disable-next-line import/no-extraneous-dependencies
import configureMockStore from 'redux-mock-store';
import { RootState } from '../src/store';

export default configureMockStore<RootState>();
