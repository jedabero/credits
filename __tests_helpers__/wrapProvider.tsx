import React, { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { Store } from '../src/store';

const wrapProvider = (component: ReactElement, store: Store) => (
  <Provider store={store}>{component}</Provider>
);

export default wrapProvider;
