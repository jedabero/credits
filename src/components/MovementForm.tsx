import {
  useState, useCallback, useMemo, useEffect,
} from 'react';
import { EntityId } from '@reduxjs/toolkit';
import { format } from 'date-fns';
import { selectMovementById, addMovement, updateMovement } from '../reducers/movements';
import { useAppSelector, useAppDispatch } from '../hooks';
import { Movement } from '../models/Movement';

type MovementFormProps = {
  cardId: EntityId,
  movementId?: EntityId,
  clearMovementId: () => void,
};

function MovementForm({ movementId, cardId, clearMovementId }: MovementFormProps) {
  const dispatch = useAppDispatch();
  const [wasSaved, setWasSaved] = useState(false);
  const id = useMemo(() => {
    if (wasSaved) setWasSaved(false);
    return movementId ?? crypto.randomUUID();
  }, [wasSaved, movementId]);
  const movement = useAppSelector((state) => selectMovementById(state, id));
  const [values, setValues] = useState<Movement>({
    id,
    cardId: movement?.cardId ?? cardId,
    authNumber: movement?.authNumber ?? '',
    date: movement?.date ?? format(new Date(), 'yyyy-MM-dd'),
    description: movement?.description ?? '',
    value: movement?.value ?? 1,
    monthlyRate: movement?.monthlyRate ?? 0,
    yearlyRate: movement?.yearlyRate ?? 0,
    quotas: movement?.quotas ?? 1,
  });
  useEffect(() => {
    setValues({
      id,
      cardId: movement?.cardId ?? cardId,
      authNumber: movement?.authNumber ?? '',
      date: movement?.date ?? format(new Date(), 'yyyy-MM-dd'),
      description: movement?.description ?? '',
      value: movement?.value ?? 1,
      monthlyRate: movement?.monthlyRate ?? 0,
      yearlyRate: movement?.yearlyRate ?? 0,
      quotas: movement?.quotas ?? 1,
    });
  }, [id, cardId, movement]);
  const setValue = useCallback(
    (event: React.ChangeEvent<HTMLInputElement & { name: keyof Movement }>) => {
      const { name, value } = event.target;
      const valueType = typeof values[name];
      if (valueType === 'number' && Number.isNaN(parseInt(value, 10))) return;
      let parsedValue: string | Date | number = value;
      if (valueType === 'number') {
        if (['monthlyRate', 'yearlyRate'].includes(name)) {
          parsedValue = parseFloat(value);
        } else {
          parsedValue = parseInt(value, 10);
        }
      }
      setValues((oldValues: Movement) => ({
        ...oldValues,
        [name]: parsedValue,
      }));
    },
    [values, setValues],
  );
  const save = useCallback(() => {
    if (movementId) {
      dispatch(updateMovement({ id: movementId, changes: values }));
      clearMovementId();
    } else {
      dispatch(addMovement(values));
    }
    setWasSaved(true);
  }, [movementId, dispatch, clearMovementId, values]);
  const cancel = useCallback(() => {
    setValues({
      id,
      cardId,
      authNumber: '',
      date: format(new Date(), 'yyyy-MM-dd'),
      description: '',
      value: 1,
      monthlyRate: 0,
      yearlyRate: 0,
      quotas: 1,
    });
    clearMovementId();
  }, [id, cardId, clearMovementId]);
  return (
    <div className="movementrow">
      <div>{(`${values.id}`).substring(0, 13)}</div>
      <input name="authNumber" value={values.authNumber} onChange={setValue} />
      <input name="date" type="date" value={values.date} onChange={setValue} />
      <input name="description" value={values.description} onChange={setValue} />
      <input name="value" type="number" value={values.value} onChange={setValue} min={1} />
      <input name="monthlyRate" type="number" value={values.monthlyRate} onChange={setValue} />
      <input name="yearlyRate" type="number" value={values.yearlyRate} onChange={setValue} />
      <input name="quotas" type="number" value={values.quotas} onChange={setValue} min={1} />
      <div>
        <button type="button" onClick={save}>{movementId ? 'Update' : 'Create'}</button>
        {movementId ? <button type="button" onClick={cancel}>Cancel</button> : null}
      </div>
    </div>
  );
}

MovementForm.defaultProps = {
  movementId: null,
};

export default MovementForm;
