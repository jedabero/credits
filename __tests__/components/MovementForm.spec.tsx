/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { create } from 'react-test-renderer';
import MovementForm from '../../src/components/MovementForm';
import wrapProvider from '../../__tests_helpers__/wrapProvider';
import mockStore from '../../__tests_helpers__/mockStore';

describe('MovementForm test', () => {
  beforeAll(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2023, 0, 1));
  });
  test('should render', () => {
    const store = mockStore({
      _persist: { version: 0, rehydrated: true },
      creditCards: {
        entities: {},
        ids: [],
      },
      movements: {
        entities: {},
        ids: [],
      },
    });
    const tree = create(
      wrapProvider(
        <MovementForm movementId="SOME ID" cardId="SOME ID" clearMovementId={() => {}} />,
        store,
      ),
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  afterAll(() => {
    jest.useRealTimers();
  });
});
