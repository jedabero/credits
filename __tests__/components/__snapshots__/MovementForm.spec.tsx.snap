// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovementForm test should render 1`] = `
<div
  className="movementrow"
>
  <div>
    SOME ID
  </div>
  <input
    name="authNumber"
    onChange={[Function]}
    value=""
  />
  <input
    name="date"
    onChange={[Function]}
    type="date"
    value="2023-01-01"
  />
  <input
    name="description"
    onChange={[Function]}
    value=""
  />
  <input
    min={1}
    name="value"
    onChange={[Function]}
    type="number"
    value={1}
  />
  <input
    name="monthlyRate"
    onChange={[Function]}
    type="number"
    value={0}
  />
  <input
    name="yearlyRate"
    onChange={[Function]}
    type="number"
    value={0}
  />
  <input
    min={1}
    name="quotas"
    onChange={[Function]}
    type="number"
    value={1}
  />
  <div>
    <button
      onClick={[Function]}
      type="button"
    >
      Update
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </div>
</div>
`;
